src/server/Connection.java

```diff
@@ -7,38 +7,44 @@
 import java.io.InputStreamReader;
 import java.io.OutputStreamWriter;
 import java.net.Socket;
+import java.util.ArrayList;
 import java.util.Arrays;
 
 public class Connection implements Runnable {
 
-    Socket socket;
-    InputStreamReader input;
-    OutputStreamWriter output;
-    Encryptor encryptor;
+    private static ArrayList<Connection> allConnections = new ArrayList<>();
+    private static int threadIDCounter = 0;
+    private Socket socket;
+    private InputStreamReader input;
+    private OutputStreamWriter output;
+    private Encryptor encryptor;
+    private final int threadID;
 
     public Connection(Socket socket) throws IOException {
         this.socket = socket;
         input = new InputStreamReader(this.socket.getInputStream());
         output = new OutputStreamWriter(this.socket.getOutputStream());
         encryptor = Encryptor.negotiateKeysServerSide(input, output);
+        this.threadID = threadIDCounter++;
+        allConnections.add(this);
```

Comment:
The encryption did work, turns out i wasn't array.copy the char arrays so they were getting encrypted multiple times.
In a different branch I implemented an online and private message system that uses the arraylists to it's advantage. I did send the name to the server similarly as an immediate send when the client connects.

Reply:
I'm not sure how I would put the sendall in the application layer, from what i understand the client's input/outputs are attached to a socket that can be read/written to by connection, which is threaded object that is instantiated for each client that connects. The way my other private message thing works is the client sends a /msg name info onto the input stream, and the connection then parses, finds the person, then fires it to the thread ID it looked up via the name. The only data transfer between the application and the server is through the socket, how could I send data to a different user without bouncing through the connection object?

Reply:
Sorry you're right that it would have to go through the connection object absolutely. Your other analysis is spot on as well. 

Part of the confusion is because I didn't realize you had some other code sitting on other branches that you hadn't made a PR for yet, so I didn't realize you had already built some other stuff based on what you'd done here.

What I'm talking about with the application layer is a server application. I'm not sure exactly how best to do it. Off the top of my head I imagine that the server application gets all of the connection objects from the Server object. Then when each connection object receives a message, it stores it in some kind of buffer Object. The server application then reads from all of it's buffers (or maybe just a single buffer that all connections add to, so each message gets handles in order), and forwards the messages on through the correct connections to other clients.

It's probably not necessary and if you have found a way that works, that's probably fine too. We're on a time crunch at the end of the day. I'll look at your other code tomorrow when I get a chance.

Reply:
Ah I'm seein the confusion here, I was thought you were talking about moving it up from connection into client, and was wondering how that would be possible.
I'd like to do it that way as well, i wrote some quick and dirty stuff that basically uses the connection as the workhorse. We will see how time goes, might have to leave it that way.